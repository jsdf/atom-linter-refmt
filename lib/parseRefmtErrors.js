'use babel';
// @flow

// work around flow libdefs not correctly typing regexp match result
function getMatchIndex(match /*: Array<string>*/) /*: number*/ {
  const unsafeMatch /*: any*/ = match;
  return unsafeMatch.index;
}

function getMatches(
  regexp /*: RegExp*/,
  input /*: string*/
) /*: Array<Array<string>>*/ {
  if (!regexp.flags.includes('g')) {
    throw new Error('expected regexp with g flag set');
  }
  const matches = [];
  let match = null;
  while ((match = regexp.exec(input))) {
    matches.push(match);
  }
  return matches;
}

/*:: type ErrorType = 'Error' | 'Warning';*/

function getErrorMessageType(typeString /*: string*/) /*: ErrorType*/ {
  switch (typeString) {
    case 'Error':
    case 'Warning':
      return typeString;
  }
  throw new Error(`unknown message type '${typeString}'`);
}

/*::
type ParsedError = {
  text: string,
  type: ErrorType,
  range: [[number, number], [number, number]],
};
*/

export function parseRefmtErrors(
  errorsText /*: ?string*/
) /*: Array<ParsedError>*/ {
  if (typeof errorsText !== 'string') {
    return [];
  }

  const matches = getMatches(
    /File "(.*?)", line (\d+), characters (\d+)-(\d+):\r?\n(Error|Warning)[:]? /g,
    errorsText
  );

  const errors = [];
  for (var i = 0; i < matches.length; i++) {
    const match = matches[i];

    const filename = match[1];
    const line = match[2];
    const colStart = match[3];
    const colEnd = match[4];
    const type = getErrorMessageType(match[5]);
    const nextMatchIndex = matches[i + 1]
      ? getMatchIndex(matches[i + 1])
      : null;
    let message = errorsText.slice(
      getMatchIndex(match) + match[0].length,
      nextMatchIndex || errorsText.length
    );
    const range = [
      [parseInt(line) - 1, parseInt(colStart)],
      [parseInt(line) - 1, parseInt(colEnd)],
    ];
    errors.push({
      text: `${type} ${message.trim()}`,
      type,
      range,
    });
  }
  return errors;
}
