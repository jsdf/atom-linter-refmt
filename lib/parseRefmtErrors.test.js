import {parseRefmtErrors} from './parseRefmtErrors';

test('single error', () => {
  expect(
    parseRefmtErrors(
      `File "", line 16, characters 34-35:
Error: 890: Expecting an expression
`
    )
  ).toMatchSnapshot();
});

test('multiple errors', () => {
  expect(
    parseRefmtErrors(
      `File "", line 1, characters 2-10:
Error: 321: Expecting an statement:
A
  b
File "", line 16, characters 34-35:
Error: 890: Expecting an expression`
    )
  ).toMatchSnapshot();
});

test('warning + error', () => {
  expect(
    parseRefmtErrors(
      `File "", line 16, characters 33-35:
Warning 2: this is not the end of a comment.
File "", line 16, characters 34-35:
Error: 890: Expecting an expression`
    )
  ).toMatchSnapshot();
});

test('no level', () => {
  expect(
    parseRefmtErrors(
      `File "", line 20, characters 7-8:
A type's name need to begin with a lower-case letter or _
`
    )
  ).toMatchSnapshot();
});
