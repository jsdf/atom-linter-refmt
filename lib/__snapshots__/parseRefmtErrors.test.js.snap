// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple errors 1`] = `
Array [
  Object {
    "range": Array [
      Array [
        0,
        2,
      ],
      Array [
        0,
        10,
      ],
    ],
    "text": "Error 321: Expecting an statement:
A
  b",
    "type": "Error",
  },
  Object {
    "range": Array [
      Array [
        15,
        34,
      ],
      Array [
        15,
        35,
      ],
    ],
    "text": "Error 890: Expecting an expression",
    "type": "Error",
  },
]
`;

exports[`no level 1`] = `
Array [
  Object {
    "range": Array [
      Array [
        19,
        7,
      ],
      Array [
        19,
        8,
      ],
    ],
    "text": "A type's name need to begin with a lower-case letter or _",
    "type": "Error",
  },
]
`;

exports[`single error 1`] = `
Array [
  Object {
    "range": Array [
      Array [
        15,
        34,
      ],
      Array [
        15,
        35,
      ],
    ],
    "text": "Error 890: Expecting an expression",
    "type": "Error",
  },
]
`;

exports[`warning + error 1`] = `
Array [
  Object {
    "range": Array [
      Array [
        15,
        33,
      ],
      Array [
        15,
        35,
      ],
    ],
    "text": "Warning 2: this is not the end of a comment.",
    "type": "Warning",
  },
  Object {
    "range": Array [
      Array [
        15,
        34,
      ],
      Array [
        15,
        35,
      ],
    ],
    "text": "Error 890: Expecting an expression",
    "type": "Error",
  },
]
`;
